### ISSUE TITLE
Using Sing-Box with ECH with Cloudflare CDN (Tunnel through CDN)

### ISSUE THREAD

**Issue by Gharib110** [net4people/bbs#431]
Hello Everyone,
Yesterday I tried to connect to my Sing-Box VPN which is behind the Cloudflare CDN with ECH Enabled. Also I had tested it with direct connection to my server successfully. it is now possible to use ECH when your Sing-Box service is behind the CDN.
just use ``` dig +short <YOUR_SNI> HTTPS ``` and retrieve the key you find ``` ech=xxxxxxxxxx ``` (Do not copy 'ech=').
You need to prepare ``` Vless or Trojan with Websocket or HTTPUpgrade ``` with sing-box v1.10.x behind Cloudflare CDN.
Then use following outbound configuration and use it along with DOH. DOH is necessary ! (Make a private DOH for yourself)

Based on my observations it does not ``` h2 ``` as ALPN and I don't know why so leave ALPN ```h3``` or ```http/1.1```

Suppose the ech key you find is 
```AEX+DQBBVgAgACBmdQmMnXF6/qYdFq+r5VhboQqZ2EzHG1gPKHRY2/k+KwAEAAEAAQASY2xvdWRmbGFyZS1lY2guY29tAAA=```
 then I divided it into two parts
```AEX+DQBBVgAgACBmdQmMnXF6/qYdFq+r5VhboQqZ2EzHG1gPKHRY2/k+KwAEAAEA```
```AQASY2xvdWRmbGFyZS1lY2guY29tAAA=```
so look at the config and put it like I used.

### JUST MY THOUGHTS and OBSERVATIONS
If an ECH config deos not work try ``` dig +short <YOUR_SNI> HTTPS  ``` again to get another key and test it again.
YOU MAY SEE different ECH keys !
I THINK keys validation EXPIRES everyday so you need to update it everyday !

I inspected the SNI of my connection to my domain with Wireshark and saw cloudflare-ech.com which indicates that this SNI will be seen by DPI systems

### The Client Outbound Config
```json
{
      "packet_encoding": "packetaddr",
      "server": "One_CLOUDFLARE_IP",
      "server_port": 443,
      "tls": {
        "alpn": [
          "h3","http/1.1"
        ],
        "ech": {
          "config": [
            "-----BEGIN ECH CONFIGS-----",
"AEX+DQBBVgAgACBmdQmMnXF6/qYdFq+r5VhboQqZ2EzHG1gPKHRY2/k+KwAEAAEA",
"AQASY2xvdWRmbGFyZS1lY2guY29tAAA=",
            "-----END ECH CONFIGS-----"
          ],
          "dynamic_record_sizing_disabled": false,
          "enabled": true,
          "pq_signature_schemes_enabled": false
        },
        "enabled": true,
        "insecure": false,
        "server_name": "YOUR_SNI",
        "utls": {
          "enabled": true,
          "fingerprint": "chrome_psk_shuffle"
        }
      },
      "transport": {
        "early_data_header_name": "Sec-WebSocket-Protocol",
        "headers": {
          "Host": "YOUR_SNI"
        },
        "max_early_data": 2048,
        "path": "/YOUR_PATH",
        "type": "ws"
      },
      "uuid": "YOUR_UUID",
      "type": "vless",
      "tag":"vless-tag-tag",
      "multiplex": {
        "enabled": true,
        "max_streams": 8,
        "padding": true,
        "protocol": "h2mux"
      }
    },
```
@irgfw
@hiddify-com

**Comment by Gharib110** [comment]
> Public DoH Servers are already blocked by Iran GFW
> GFW can remove ECH key from unencrypted DNS responses
> Browsers with ECH support also try ECH and then fallbacks to unencrypted SNI
> so GFW can safely block all ECH TLS requests, they will do it soon or later
> 
> But for now, the matter is that if that ECH key expires or not, if it expires then proxy client must support getting the ECH key from a custom encrypted DNS server before connecting to the server
> 

Yes Public DOH are blocked. Iranians have provided many sources to run a private DOH. So I have private DOH for myself.
I have developed a method to retrieve ech key every 1 hour safely and make the client configuration.

**Comment by Gharib110** [comment]
> Public DoH Servers are already blocked by Iran GFW
> GFW can remove ECH key from unencrypted DNS responses
> Browsers with ECH support also try ECH and then fallbacks to unencrypted SNI
> so GFW can safely block all ECH TLS requests, they will do it soon or later
> 
> But for now, the matter is that if that ECH key expires or not, if it expires then proxy client must support getting the ECH key from a custom encrypted DNS server before connecting to the server
> 

Please describe manifestly how they can block ech tls requests? by blocking the camouflage SNI ```cloudflare-ech.com``` ? or other methods?
if they block the SNI, you can use another SNI.

**Comment by kovalensky** [comment]
> Please describe manifestly how they can block ech tls requests? by blocking the camouflage SNI `cloudflare-ech.com` ? or other methods?

By [specification](https://www.ietf.org/archive/id/draft-ietf-tls-esni-08.html#name-do-not-stick-out) clienthello handshake has a ECH protocol identificitation.

**Comment by Gharib110** [comment]
> > Please describe manifestly how they can block ech tls requests? by blocking the camouflage SNI `cloudflare-ech.com` ? or other methods?
> 
> By [specification](https://www.ietf.org/archive/id/draft-ietf-tls-esni-08.html#name-do-not-stick-out) clienthello handshake has an ECH protocol identificitation.

Thank you for the reference.
I have noticed you're Russian and saw another issue discussing about ech that is blocked partially based on their claims. Someone said the ECH with ``cloudflare-ech.com``` is blocked but with another SNI like (you can set another outer SNI and generate keys with sing-box and send traffic directly to your server ) is not blocked. Do you have any thoughts about it?
Thanks
